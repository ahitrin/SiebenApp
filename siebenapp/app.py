#!/usr/bin/env python3
# coding: utf-8
import sys
from argparse import ArgumentParser
from os.path import dirname, join, realpath

from PyQt5.QtCore import pyqtSignal, Qt, QRect
from PyQt5.QtGui import QPainter, QPen
from PyQt5.QtWidgets import QApplication, QMainWindow, QWidget, QGridLayout
from PyQt5.uic import loadUi

from siebenapp.goaltree import Edge
from siebenapp.render import Renderer
from siebenapp.system import save, load, DEFAULT_DB, split_long
from siebenapp.ui.goalwidget import Ui_GoalBody


class GoalWidget(QWidget, Ui_GoalBody):
    clicked = pyqtSignal()

    def __init__(self):
        super().__init__()
        self._click_in_progress = False
        self.setupUi(self)
        self.is_real = True
        self.widget_id = None

    def setup_data(self, number, attributes):
        self.widget_id = number
        self.label_goal_name.setText(split_long(attributes['name']))
        self.check_open.setVisible(attributes['switchable'])
        self.is_real = isinstance(number, int)
        selection = attributes['select']
        if selection == 'select':
            self.setStyleSheet('background-color:#808080;')
        elif selection == 'prev':
            self.setStyleSheet('background-color:#C0C0C0;')
        if self.is_real:
            frame_color = 'red' if attributes['open'] else 'green'
            border = 2 if attributes['switchable'] else 1
            self.frame.setStyleSheet('.QFrame{ border: %dpx solid %s }' % (border, frame_color))
            self.label_number.setText(str(number))
        else:
            self.setStyleSheet('color: #EEEEEE; border: #EEEEEE')

    def mousePressEvent(self, event):                           # pylint: disable=unused-argument
        self._click_in_progress = True

    def mouseReleaseEvent(self, event):                         # pylint: disable=unused-argument
        if self._click_in_progress:
            self.clicked.emit()
        self._click_in_progress = False

    def top_point(self):
        rect = self.geometry()
        return (rect.topLeft() + rect.topRight()) / 2

    def bottom_point(self):
        rect = self.geometry()
        return (rect.bottomLeft() + rect.bottomRight()) / 2


class CentralWidget(QWidget):
    EDGE_PENS = {
        Edge.BLOCKER: QPen(Qt.black, 1, Qt.DashLine),
        Edge.PARENT: QPen(Qt.black, 1, Qt.SolidLine),
    }

    def __init__(self):
        super().__init__()
        self.setGeometry(QRect(0, 0, 576, 273))
        self.setObjectName('scrollAreaWidgetContents')
        layout = QGridLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        self.setLayout(layout)
        self.dependencies = dict()

    def setDependencies(self, new_dependencies):
        self.dependencies = new_dependencies

    def paintEvent(self, event):                                    # pylint: disable=unused-argument
        painter = QPainter(self)

        widgets = {w.widget_id: (w.top_point(), w.bottom_point(), w.is_real)
                   for w in self.children() if isinstance(w, GoalWidget)}
        for widget_id, points in widgets.items():
            line_start = points[0]
            for edge in self.dependencies[widget_id]:
                line_end = widgets[edge[0]][1]
                painter.setPen(self.EDGE_PENS[edge[1]])
                painter.drawLine(line_start, line_end)
            if not points[2]:
                style = max([e[1] for e in self.dependencies[widget_id]])
                painter.setPen(self.EDGE_PENS[style])
                painter.drawLine(points[0], points[1])


class SiebenApp(QMainWindow):
    refresh = pyqtSignal()
    quit_app = pyqtSignal()

    def __init__(self, db, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.refresh.connect(self.save_and_render)
        self.quit_app.connect(QApplication.instance().quit)
        self.db = db
        self.goals = load(db, self.show_user_message)
        self.force_refresh = True

    def setup(self):
        self.action_Hotkeys.triggered.connect(self.hotkeys.show)
        self.action_About.triggered.connect(self.about.show)
        # Re-creation of scrollAreaWidgetContents looks like dirty hack,
        # but at the current moment I haven't found a better solution.
        # Widget creation in __init__ does not work: lines disappear.
        # Also we have to disable pylint warning in order to make build green.
        self.scrollAreaWidgetContents = CentralWidget()         # pylint: disable=attribute-defined-outside-init
        # End of 'looks like dirty hack'
        self.scrollArea.setWidget(self.scrollAreaWidgetContents)
        self._update_title()
        self.refresh.emit()

    def _update_title(self):
        self.setWindowTitle('{0} ({1} view) - SiebenApp'.format(self.db, self.goals.view))

    def close_goal(self, goal_id):
        def inner():
            self.goals.select(goal_id)
            self.goals.toggle_close()
            self.refresh.emit()
        return inner

    def save_and_render(self):
        if not self.goals.events and not self.force_refresh:
            return
        self.force_refresh = False
        self.statusBar().clearMessage()
        save(self.goals, self.db)
        for child in self.scrollAreaWidgetContents.children():
            if isinstance(child, GoalWidget):
                child.deleteLater()
        graph = Renderer(self.goals).build()
        if 'setDependencies' in dir(self.scrollAreaWidgetContents):
            self.scrollAreaWidgetContents.setDependencies({g: graph[g]['edge'] for g in graph})
        for goal_id, attributes in graph.items():
            widget = GoalWidget()
            self.scrollAreaWidgetContents.layout().addWidget(widget, attributes['row'], attributes['col'])
            widget.setup_data(goal_id, attributes)
            if isinstance(goal_id, int):
                widget.clicked.connect(self.select_number(goal_id))
                widget.check_open.clicked.connect(self.close_goal(goal_id))
        self.scrollAreaWidgetContents.update()

    def keyPressEvent(self, event):
        key_handlers = {
            Qt.Key_1: self.select_number(1),
            Qt.Key_2: self.select_number(2),
            Qt.Key_3: self.select_number(3),
            Qt.Key_4: self.select_number(4),
            Qt.Key_5: self.select_number(5),
            Qt.Key_6: self.select_number(6),
            Qt.Key_7: self.select_number(7),
            Qt.Key_8: self.select_number(8),
            Qt.Key_9: self.select_number(9),
            Qt.Key_0: self.select_number(0),
            Qt.Key_A: self.start_edit('Add new goal', self.goals.add),
            Qt.Key_C: self.with_refresh(self.goals.toggle_close),
            Qt.Key_D: self.with_refresh(self.goals.delete),
            Qt.Key_I: self.start_edit('Insert new goal', self.goals.insert),
            Qt.Key_K: self.with_refresh(self.goals.toggle_link, edge_type=Edge.PARENT),
            Qt.Key_L: self.with_refresh(self.goals.toggle_link),
            Qt.Key_Q: self.quit_app.emit,
            Qt.Key_R: self.start_edit('Rename goal', self.goals.rename, self._current_goal_label),
            Qt.Key_V: self.toggle_view,
            Qt.Key_Z: self.toggle_zoom,
            Qt.Key_Escape: self.cancel_edit,
            Qt.Key_Space: self.with_refresh(self.goals.hold_select),
            Qt.Key_Slash: self.show_keys_help,
        }
        if event.key() in key_handlers:
            key_handlers[event.key()]()
        else:
            super().keyPressEvent(event)

    def start_edit(self, label, fn, pre_fn=None):
        def inner():
            self.dockWidget.setWindowTitle(label)
            self.input.setEnabled(True)
            self.input.setFocus(True)
            if pre_fn is not None:
                self.input.setText(pre_fn())
            self.input.returnPressed.connect(self.finish_edit(fn))
            self.cancel.setEnabled(True)
            self.cancel.clicked.connect(self.cancel_edit)
        return inner

    def finish_edit(self, fn):
        def inner():
            self.dockWidget.setWindowTitle('')
            self.input.returnPressed.disconnect()
            fn(self.input.text())
            self.input.setEnabled(False)
            self.input.setText('')
            self.cancel.setEnabled(False)
            self.cancel.clicked.disconnect()
            self.refresh.emit()
        return inner

    def cancel_edit(self):
        self.dockWidget.setWindowTitle('')
        self.input.setEnabled(False)
        self.input.setText('')
        self.cancel.setEnabled(False)
        try:
            self.input.returnPressed.disconnect()
            self.cancel.clicked.disconnect()
        except TypeError:
            pass

    def _current_goal_label(self):
        data = self.goals.q(keys='name,select').values()
        return [x['name'] for x in data if x['select'] == 'select'].pop()

    def with_refresh(self, fn, *args, **kwargs):
        def inner():
            fn(*args, **kwargs)
            self.refresh.emit()
        return inner

    def select_number(self, num):
        def inner():
            self.goals.select(num)
            self.refresh.emit()
        return inner

    def toggle_view(self):
        self.force_refresh = True
        self.goals.next_view()
        self._update_title()
        self.refresh.emit()

    def toggle_zoom(self):
        self.force_refresh = True
        self.goals.toggle_zoom()
        self.refresh.emit()

    def show_keys_help(self):
        self.action_Hotkeys.trigger()

    def show_user_message(self, message):
        self.statusBar().showMessage(message, 10000)


def main(root_script):
    parser = ArgumentParser()
    parser.add_argument('db', nargs='?', default=DEFAULT_DB,
                        help='Path to the database file (sieben.db by default)')
    args = parser.parse_args()
    app = QApplication(sys.argv)
    root = dirname(realpath(root_script))
    w = loadUi(join(root, 'ui', 'main.ui'), SiebenApp(args.db))
    w.about = loadUi(join(root, 'ui', 'about.ui'))
    w.hotkeys = loadUi(join(root, 'ui', 'hotkeys.ui'))
    w.setup()
    w.showMaximized()
    sys.exit(app.exec_())
