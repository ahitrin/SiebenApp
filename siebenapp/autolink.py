from dataclasses import dataclass
from typing import Dict, List, Union, Tuple, Optional, Set

from siebenapp.domain import (
    Graph,
    Command,
    EdgeType,
    ToggleClose,
    Add,
    ToggleLink,
    Insert,
    Rename,
    Delete,
    GoalId,
    RenderResult,
    RenderRow,
)
from siebenapp.goaltree import Goals


@dataclass(frozen=True)
class ToggleAutoLink(Command):
    keyword: str


AutoLinkData = List[Tuple[int, str]]


class AutoLink(Graph):
    def __init__(self, goals: Graph, data: Optional[AutoLinkData] = None):
        super(AutoLink, self).__init__(goals)
        self.keywords: Dict[str, int] = {}
        self.back_kw: Dict[int, str] = {}
        if data:
            for goal_id, keyword in data:
                self.keywords[keyword] = goal_id
                self.back_kw[goal_id] = keyword

    def accept_ToggleAutoLink(self, command: ToggleAutoLink) -> None:
        selected_id: int = self.settings("selection")
        if selected_id in self.goaltree.closed:
            self.error("Autolink cannot be set for closed goals")
            return
        if selected_id == Goals.ROOT_ID:
            self.error("Autolink cannot be set for the root goal")
            return
        keyword: str = command.keyword.lower().strip()
        if selected_id in self.back_kw:
            self.keywords.pop(self.back_kw[selected_id])
            self.back_kw.pop(selected_id)
            self.events().append(("remove_autolink", selected_id))
        if keyword in self.keywords:
            old_id: int = self.keywords.pop(keyword)
            self.back_kw.pop(old_id)
            self.events().append(("remove_autolink", old_id))
        if not keyword:
            # empty keyword? exit right now
            return
        self.keywords[keyword] = selected_id
        self.back_kw[selected_id] = keyword
        self.events().append(("add_autolink", selected_id, keyword))

    def accept_ToggleClose(self, command: ToggleClose) -> None:
        selected_id: int = self.settings("selection")
        if selected_id in self.back_kw:
            self.keywords.pop(self.back_kw[selected_id])
            self.back_kw.pop(selected_id)
            self.events().append(("remove_autolink", selected_id))
        self.goaltree.accept(command)

    def accept_Add(self, command: Add) -> None:
        self._autolink_new_goal(command)

    def accept_Insert(self, command: Insert) -> None:
        self._autolink_new_goal(command)

    def _autolink_new_goal(self, command: Union[Add, Insert]) -> None:
        matching: List[int] = self._find_matching_goals(command.name)
        ids_before: Set[int] = set(self.goaltree.goals.keys())
        self.goaltree.accept(command)
        ids_after: Set[int] = set(self.goaltree.goals.keys())
        ids_diff: Set[int] = ids_after.difference(ids_before)
        if ids_diff:
            added_id: int = ids_diff.pop()
            self._make_links(matching, added_id)

    def accept_Rename(self, command: Rename) -> None:
        matching: List[int] = self._find_matching_goals(command.new_name)
        self.goaltree.accept(command)
        selected_id: int = command.goal_id or self.settings("selection")
        self._make_links(matching, selected_id)

    def accept_Delete(self, command: Delete) -> None:
        selected_id: int = command.goal_id or self.settings("selection")
        edges: Dict[GoalId, List[Tuple[GoalId, EdgeType]]] = {
            row.goal_id: row.edges for row in self.goaltree.q().rows
        }
        goals_to_check: List[int] = [selected_id]
        while goals_to_check:
            goal_id: int = goals_to_check.pop()
            goals_to_check.extend(
                e[0]
                for e in edges[goal_id]
                if e[1] == EdgeType.PARENT and isinstance(e[0], int)
            )
            if goal_id in self.back_kw:
                added_kw: str = self.back_kw.pop(goal_id)
                self.keywords.pop(added_kw)
                self.events().append(("remove_autolink", goal_id))
        self.goaltree.accept(command)

    def _find_matching_goals(self, text: str) -> List[int]:
        return [goal_id for kw, goal_id in self.keywords.items() if kw in text.lower()]

    def _make_links(self, matching_goals: List[int], target_goal: int) -> None:
        if not matching_goals:
            return
        self_children: Dict[GoalId, List[GoalId]] = {
            row.goal_id: [e[0] for e in row.edges] for row in self.goaltree.q().rows
        }
        for add_to in matching_goals:
            if target_goal not in self_children[add_to]:
                self.goaltree.accept(ToggleLink(add_to, target_goal, EdgeType.BLOCKER))

    def q(self) -> RenderResult:
        rows: List[RenderRow] = [
            RenderRow(
                row.goal_id,
                row.raw_id,
                row.name,
                row.is_open,
                row.is_switchable,
                row.select,
                self._add_pseudo_goals(row.edges),
            )
            for row in self.goaltree.q().rows
        ]
        fake_rows: List[RenderRow] = [
            RenderRow(
                AutoLink.fake_id(goal_id),
                -1,
                f"Autolink: '{keyword}'",
                True,
                False,
                None,
                [(goal_id, EdgeType.PARENT)],
            )
            for keyword, goal_id in self.keywords.items()
        ]
        return RenderResult(rows=rows + fake_rows)

    def _add_pseudo_goals(
        self, edges: List[Tuple[GoalId, EdgeType]]
    ) -> List[Tuple[GoalId, EdgeType]]:
        return [
            e if e[0] not in self.keywords.values() else (AutoLink.fake_id(e[0]), e[1])
            for e in edges
        ]

    @staticmethod
    def fake_id(goal_id: GoalId) -> GoalId:
        return -(goal_id + 10) if isinstance(goal_id, int) else goal_id

    @staticmethod
    def export(goals: "AutoLink") -> AutoLinkData:
        return [(goal_id, kw) for goal_id, kw in goals.back_kw.items()]
