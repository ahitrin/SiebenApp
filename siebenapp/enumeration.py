import math
from typing import List, Tuple, Any, Iterable, Set

from siebenapp.domain import Graph, Select, GoalId, RenderResult, RenderRow


class BidirectionalIndex:
    NOT_FOUND = -2

    def __init__(self, goals: Iterable[GoalId]):
        self.m = {
            g: i + 1
            for i, g in enumerate(
                sorted(g for g in goals if isinstance(g, int) and g > 0)
            )
        }
        self.length = len(self.m)

    def forward(self, goal_id: GoalId) -> int:
        if isinstance(goal_id, str):
            return BidirectionalIndex.NOT_FOUND
        if goal_id < 0:
            return goal_id
        goal_id = self.m[goal_id]
        new_id = goal_id % 10
        if self.length > 10:
            new_id += 10 * ((goal_id - 1) // 10 + 1)
        if self.length > 90:
            new_id += 100 * ((goal_id - 1) // 100 + 1)
        if self.length > 900:
            new_id += 1000 * ((goal_id - 1) // 1000 + 1)
        return new_id

    def backward(self, goal_id: int) -> int:
        possible_selections: List[int] = [
            g for g in self.m if self.forward(g) == goal_id
        ]
        if len(possible_selections) == 1:
            return possible_selections[0]
        return BidirectionalIndex.NOT_FOUND


class Enumeration(Graph):
    def __init__(self, goaltree: Graph) -> None:
        super().__init__(goaltree)
        self.selection_cache: List[int] = []

    def _id_mapping(self) -> Tuple[RenderResult, BidirectionalIndex]:
        render_result = self.goaltree.q()
        return render_result, BidirectionalIndex(
            [r.goal_id for r in render_result.rows]
        )

    def settings(self, key: str) -> Any:
        if key == "root":
            _, index = self._id_mapping()
            return index.forward(self.goaltree.settings("root"))
        return self.goaltree.settings(key)

    def q(self) -> RenderResult:
        render_result, index = self._id_mapping()
        rows: List[RenderRow] = [
            RenderRow(
                index.forward(row.goal_id),
                row.raw_id,
                row.name,
                row.is_open,
                row.is_switchable,
                [(index.forward(e[0]), e[1]) for e in row.edges],
            )
            for row in render_result.rows
        ]
        new_select = (
            index.forward(render_result.select[0]),
            index.forward(render_result.select[1]),
        )
        return RenderResult(rows=rows, select=new_select)

    def accept_Select(self, command: Select):
        render_result, index = self._id_mapping()
        goals: Set[GoalId] = {row.goal_id for row in render_result.rows}
        if (goal_id := command.goal_id) >= 10:
            self.selection_cache = []
        if self.selection_cache:
            goal_id = 10 * self.selection_cache.pop() + goal_id
            if goal_id > max(index.forward(k) for k in goals if isinstance(k, int)):
                goal_id %= int(pow(10, int(math.log(goal_id, 10))))
        if (original_id := index.backward(goal_id)) != BidirectionalIndex.NOT_FOUND:
            self.goaltree.accept(Select(original_id))
            self.selection_cache = []
        else:
            self.selection_cache.append(goal_id)
