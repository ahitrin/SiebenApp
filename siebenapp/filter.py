import collections
from dataclasses import dataclass
from typing import Dict, Any

from siebenapp.domain import Graph, Command, with_key


@dataclass(frozen=True)
class FilterBy(Command):
    """Activate or deactivate filtering"""

    pattern: str


class FilterView(Graph):
    def __init__(self, goaltree: Graph):
        super().__init__(goaltree)
        self.pattern = ""

    def events(self) -> collections.deque:
        return self.goaltree.events()

    def verify(self) -> bool:
        return self.goaltree.verify()

    def settings(self, key: str) -> int:
        return self.goaltree.settings(key)

    def accept_FilterBy(self, event: FilterBy):
        self.pattern = event.pattern.lower()

    @with_key("name")
    @with_key("select")
    def q(self, keys: str = "name") -> Dict[int, Any]:
        unfiltered = self.goaltree.q(keys)
        return self._filter(unfiltered, keys) if self.pattern else unfiltered

    def _filter(self, unfiltered, keys):
        accepted_nodes = [
            goal_id
            for goal_id, attrs in unfiltered.items()
            if self.pattern in attrs["name"].lower()
            or attrs.get("select", None) is not None
        ]
        filtered: Dict[int, Any] = {}
        for goal_id, attrs in unfiltered.items():
            if goal_id in accepted_nodes:
                if "edge" in keys:
                    attrs["edge"] = [e for e in attrs["edge"] if e[0] in accepted_nodes]
                filtered[goal_id] = attrs
        return filtered
