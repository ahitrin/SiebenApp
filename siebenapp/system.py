# coding: utf-8
import sqlite3
from html import escape
from os import path
from typing import Callable, List, Dict, Set

from siebenapp.domain import EdgeType, Graph
from siebenapp.enumeration import Enumeration
from siebenapp.goaltree import Goals, GoalsData, EdgesData, OptionsData
from siebenapp.layers import wrap_with_views
from siebenapp.zoom import Zoom

MIGRATIONS = [
    # 0
    [
        "create table migrations (version integer)",
        "insert into migrations values (-1)",
    ],
    # 1
    [
        """create table goals (
            goal_id integer primary key,
            name string,
            open boolean
        )""",
        """create table edges (
            parent integer,
            child integer,
            foreign key(parent) references goals(goal_id),
            foreign key(child) references goals(goal_id)
        )""",
        """create table selection (
            name string,
            goal integer,
            foreign key(goal) references goals(goal_id)
        )""",
    ],
    # 2
    [
        # change type of goals.name: string -> text
        """alter table goals rename to old_goals""",
        """create table goals (
            goal_id integer primary key,
            name text,
            open boolean
        )""",
        """insert into goals (goal_id, name, open)
           select goal_id, name, open from old_goals""",
        """drop table old_goals""",
        # change type of selection.name: string -> text
        """alter table selection rename to old_selection""",
        """create table selection (
            name text,
            goal integer,
            foreign key(goal) references goals(goal_id)
        )""",
        """insert into selection (name, goal)
           select name, goal from old_selection""",
        """drop table old_selection""",
    ],
    # 3
    [
        "alter table selection rename to settings",
    ],
    # 4
    [
        "alter table edges add column reltype integer not null default 1",
    ],
    # 5
    [
        """create table zoom (
            level integer primary key,
            goal integer,
            foreign key(goal) references goals(goal_id)
        )""",
        """insert into zoom (level, goal) select 1, 1 from settings
           where name = 'zoom' """,
        """insert into zoom (level, goal) select 2, goal from settings
           where name = 'zoom' and goal <> 1""",
        "delete from settings where name = 'zoom'",
    ],
    # 6
    [
        """create table new_edges (
            parent integer not null,
            child integer not null,
            reltype integer not null default 1,
            primary key(parent, child),
            foreign key(parent) references goals(goal_id),
            foreign key(child) references goals(goal_id)
        )""",
        "insert into new_edges select parent, child, 1 from edges",
        """insert or replace into new_edges
        select min(parent), child, 2 from edges where reltype=2 group by child""",
        "delete from edges",
        "insert into edges select * from new_edges",
    ],
    # 7
    [
        "drop table new_edges",
    ],
]


def save(goals: Graph, filename: str) -> None:
    if path.isfile(filename):
        connection = sqlite3.connect(filename)
        run_migrations(connection)
        save_updates(goals, connection)
    else:
        connection = sqlite3.connect(filename)
        run_migrations(connection)
        root_goals = goals
        while not isinstance(root_goals, Goals):
            root_goals = root_goals.goaltree
        goals_export, edges_export, select_export = Goals.export(root_goals)
        zoom_goals = goals
        while not isinstance(zoom_goals, Zoom):
            zoom_goals = zoom_goals.goaltree
        zoom_export = Zoom.export(zoom_goals)
        cur = connection.cursor()
        cur.executemany("insert into goals values (?,?,?)", goals_export)
        cur.executemany("insert into edges values (?,?,?)", edges_export)
        cur.executemany("insert into settings values (?,?)", select_export)
        cur.executemany("insert into zoom values (?, ?)", zoom_export)
        root_goals._events.clear()
        connection.commit()

    connection.close()


def save_updates(goals: Graph, connection: sqlite3.Connection) -> None:
    actions = {
        "add": ["insert into goals values (?,?,?)"],
        "toggle_close": ["update goals set open=? where goal_id=?"],
        "rename": ["update goals set name=? where goal_id=?"],
        "link": ["insert into edges values (?,?,?)"],
        "unlink": ["delete from edges where parent=? and child=? and reltype=?"],
        "select": [
            'delete from settings where name="selection"',
            'insert into settings values ("selection", ?)',
        ],
        "hold_select": [
            'delete from settings where name="previous_selection"',
            'insert into settings values ("previous_selection", ?)',
        ],
        "delete": [
            "delete from goals where goal_id=?",
            "delete from edges where child=?",
            "delete from edges where parent=?",
        ],
        "zoom": ["insert into zoom values (?, ?)"],
        "unzoom": ["delete from zoom where goal=?"],
    }
    cur = connection.cursor()
    while goals.events():
        event = goals.events().popleft()
        if event[0] in actions:
            for query in actions[event[0]]:
                if "?" in query:
                    cur.execute(query, event[1:])
                else:
                    cur.execute(query)
    connection.commit()


def load(filename: str, message_fn: Callable[[str], None] = None) -> Enumeration:
    if path.isfile(filename):
        connection = sqlite3.connect(filename)
        run_migrations(connection)
        cur = connection.cursor()
        names = list(cur.execute("select * from goals"))
        edges = list(cur.execute("select parent, child, reltype from edges"))
        settings = list(cur.execute("select * from settings"))
        zoom_data = list(cur.execute("select * from zoom"))
        cur.close()
        goals = Goals.build(names, edges, settings, message_fn)
        zoom = Zoom.build(goals, zoom_data)
    else:
        goals = Goals("Rename me", message_fn)
        zoom = Zoom(goals)
    return Enumeration(wrap_with_views(zoom))


def run_migrations(
    conn: sqlite3.Connection, migrations: List[List[str]] = None
) -> None:
    if migrations is None:
        migrations = MIGRATIONS
    cur = conn.cursor()
    try:
        cur.execute("select version from migrations")
        current_version = cur.fetchone()[0]
    except sqlite3.OperationalError:
        current_version = -1
    for num, migration in list(enumerate(migrations))[current_version + 1 :]:
        for query in migration:
            cur.execute(query)
        cur.execute("update migrations set version=?", (num,))
        conn.commit()


def split_long(line: str) -> str:
    margin = 20
    parts = []
    space_position = line.find(" ", margin)
    while space_position > 0:
        part, line = line[:space_position], line[(space_position + 1) :]
        parts.append(part)
        space_position = line.find(" ", margin)
    parts.append(line)
    return "\n".join(parts)


def _format_name(num: int, goal: Dict[str, str]) -> str:
    goal_name = escape(goal["name"])
    label = f'"{num}: {goal_name}"' if num >= 0 else f'"{goal_name}"'
    return split_long(label)


def dot_export(goals):
    data = goals.q(keys="open,name,edge,switchable")
    lines = []
    for num in sorted(data.keys()):
        goal = data[num]
        style = None
        if goal["switchable"] and goal["open"]:
            style = "bold"
        attributes = {
            "label": _format_name(num, goal),
            "color": "red" if goal["open"] else "green",
        }
        if style is not None:
            attributes["style"] = style
        attributes_str = ", ".join(
            f"{k}={attributes[k]}"
            for k in ["label", "color", "style", "fillcolor"]
            if k in attributes and attributes[k]
        )
        lines.append(f"{num} [{attributes_str}];")
    for num in sorted(data.keys()):
        for edge in data[num]["edge"]:
            color = "black" if data[edge[0]]["open"] else "gray"
            line_attrs = f"color={color}"
            if edge[1] == EdgeType.BLOCKER:
                line_attrs += ", style=dashed"
            lines.append(f"{edge[0]} -> {num} [{line_attrs}];")
    body = "\n".join(lines)
    return f"digraph g {{\nnode [shape=box];\n{body}\n}}"


def extract_subtree(source_goals: Graph, goal_id: int) -> Graph:
    while not isinstance(source_goals, Goals):
        source_goals = source_goals.goaltree
    source_data = source_goals.q(keys="name,edge,open")
    assert goal_id in source_data.keys(), f"Cannot find goal with id {goal_id}"
    target_goals: Set[int] = set()
    goals_to_add: Set[int] = {goal_id}
    goals_data: GoalsData = []
    edges_data: EdgesData = []
    options_data: OptionsData = []
    while goals_to_add:
        goal = goals_to_add.pop()
        attrs = source_data[goal]
        target_goals.add(goal)
        goals_data.append((goal, attrs["name"], attrs["open"]))
        edges_data.extend((goal, target_, type_) for target_, type_ in attrs["edge"])
        goals_to_add.update(
            set(
                edge[0]
                for edge in attrs["edge"]
                if edge[1] == EdgeType.PARENT and edge[0] not in target_goals
            )
        )
    edges_data = [edge for edge in edges_data if edge[1] in target_goals]
    remap = {
        old: idx + 2
        for idx, old in enumerate(g for g in sorted(target_goals) if g != goal_id)
    }
    remap[goal_id] = Goals.ROOT_ID
    goals_data = [
        (remap[goal_id], name, is_open) for goal_id, name, is_open in goals_data
    ]
    edges_data = [
        (remap[source], remap[target], e_type) for source, target, e_type in edges_data
    ]
    return Zoom.build(Goals.build(goals_data, edges_data, options_data), [])
