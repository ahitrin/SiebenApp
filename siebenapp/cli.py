import sys
from argparse import ArgumentParser
from operator import itemgetter

from siebenapp.render import Renderer
from siebenapp.domain import (
    ToggleClose,
    Delete,
    HoldSelect,
    ToggleLink,
    EdgeType,
    Select,
    Add,
    Insert,
    Rename,
    Graph,
)
from siebenapp.progress_view import ToggleProgress
from siebenapp.filter_view import FilterBy
from siebenapp.open_view import ToggleOpenView
from siebenapp.switchable_view import ToggleSwitchableView
from siebenapp.system import save, load
from siebenapp.zoom import ToggleZoom

USER_MESSAGE = ""


class IO:
    """Simple wrapper for all input/output interactions"""

    def write(self, text: str, *args) -> None:
        raise NotImplementedError

    def read(self) -> str:
        raise NotImplementedError


class ConsoleIO(IO):
    def __init__(self, prompt):
        super().__init__()
        sys.stdin.reconfigure(encoding="utf-8")
        self.prompt = prompt

    def write(self, text: str, *args) -> None:
        print(text, *args)

    def read(self) -> str:
        return input(self.prompt)


def update_message(message=""):
    global USER_MESSAGE
    USER_MESSAGE = message


def fmt(goal_id, goal_vars, id_width):
    if goal_id < 0:
        goal_id = " " * id_width
    name = goal_vars["name"]
    op = " " if goal_vars["open"] else "x"
    status = f"[{op}]" if goal_vars["switchable"] else f" {op} "
    selection = " "
    if goal_vars["select"] == "select":
        selection = ">"
    elif goal_vars["select"] == "prev":
        selection = "_"
    children = ""
    if goal_vars["edge"]:
        child_list = ",".join(str(e[0]) for e in goal_vars["edge"])
        children = f" [{child_list}]"
    return f"{goal_id}{status}{selection}{name}{children}"


def build_actions(command):
    simple_commands = {
        "c": ToggleClose(),
        "d": Delete(),
        "h": HoldSelect(),
        "k": ToggleLink(edge_type=EdgeType.PARENT),
        "l": ToggleLink(),
        "n": ToggleOpenView(),
        "p": ToggleProgress(),
        "t": ToggleSwitchableView(),
        "z": ToggleZoom(),
    }
    if command and all(c in "1234567890" for c in command):
        return [Select(int(c)) for c in command]
    if command.startswith("a "):
        return [Add(command[2:])]
    if command.startswith("i "):
        return [Insert(command[2:])]
    if command.startswith("r "):
        return [Rename(command[2:])]
    if command.startswith("f"):
        return [FilterBy(command[1:].lstrip())]
    if command in simple_commands:
        return [simple_commands[command]]
    return []


def loop(io: IO, goals: Graph, db_name: str):
    cmd = ""
    while cmd != "q":
        render_result = Renderer(goals, 100).build()
        rows = render_result.graph
        index = sorted(
            [
                (goal_id, goal_vars["row"], goal_vars["col"])
                for goal_id, goal_vars in rows.items()
                if isinstance(goal_id, int)
            ],
            key=itemgetter(1, 2),
            reverse=True,
        )
        id_width = len(str(max(g for g in rows.keys() if isinstance(g, int))))
        for item in index:
            goal_id = item[0]
            goal_vars = rows[goal_id]
            io.write(fmt(goal_id, goal_vars, id_width))
        if USER_MESSAGE:
            io.write(USER_MESSAGE)
        update_message()
        try:
            cmd = io.read().strip()
        except EOFError:
            break
        actions = build_actions(cmd)
        if actions:
            goals.accept_all(*actions)
        save(goals, db_name)


def main():
    io = ConsoleIO(prompt="> ")
    parser = ArgumentParser()
    parser.add_argument(
        "db",
        nargs="?",
        default="sieben.db",
        help="Path to the database file",
    )
    args = parser.parse_args()
    goals = load(args.db, update_message)
    loop(io, goals, args.db)
