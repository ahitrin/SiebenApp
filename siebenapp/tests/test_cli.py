from typing import List

from approvaltests import verify, Options  # type: ignore
from approvaltests.reporters.generic_diff_reporter import GenericDiffReporter  # type: ignore

from siebenapp.cli import IO, update_message, loop
from siebenapp.system import load


class DummyIO(IO):
    def __init__(self, commands: List[str], log: List[str]):
        super().__init__()
        self.commands = commands
        self.log = log

    def write(self, text: str, *args) -> None:
        self.log.append(" ".join([text, *args]))

    def read(self) -> str:
        if self.commands:
            command = self.commands.pop(0)
            self.log.append(f"> {command}\n")
            return command
        self.log.append("*USER BREAK*")
        raise EOFError("Commands list is empty")


def test_simple_scenario():
    log = []
    commands = [
        "r Approval testing",
        "a Measure coverage",
        "2",
        "a Build simple test",
        "3",
        "c",
    ]
    io = DummyIO(commands, log)
    db_name = ":memory:"
    goals = load(db_name, update_message)
    loop(io, goals, db_name)
    verify(
        "\n".join(log),
        options=Options().with_reporter(GenericDiffReporter.create("/usr/bin/diff")),
    )
