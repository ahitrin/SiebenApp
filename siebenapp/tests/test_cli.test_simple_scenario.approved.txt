1[ ]>Rename me
> r Approval testing

1[ ]>Approval testing
> a Measure coverage

1   >Approval testing [2]
2[ ] Measure coverage
> 2

1   _Approval testing [2]
2[ ]>Measure coverage
> a Build simple test

1   _Approval testing [2]
2   >Measure coverage [3]
3[ ] Build simple test
> 3

1   _Approval testing [2]
2    Measure coverage [3]
3[ ]>Build simple test
> c

1    Approval testing [2]
2[ ]>Measure coverage
*USER BREAK*
